docs: ADR-0002 Go Migration (#109)
RedHatProductSecurity/lola: docs/dev-guide/design/go-migration.md

# Go Migration — Implementation Design

Paired with [ADR-0002: Go Migration](../../adr/0002-go-migration.md).

## Migration Timeline

```mermaid
gantt
    title Lola Go Migration
    dateFormat YYYY-MM
    section Preparation
        Freeze Python (v1.x-python-final tag)   :2026-05, 1w
        Scaffold Go project structure (ADR-0003) :2026-05, 1w
    section Core
        CLI commands (mod, install, update, repo) :2026-05, 4w
        Module model and auto-discovery           :2026-05, 2w
        Source handlers (git, zip, tar, folder)    :2026-06, 2w
        Target generators (claude-code, cursor, gemini, openclaw, opencode) :2026-06, 3w
        Installation registry and sync             :2026-06, 2w
    section Integration
        OCI source handler (skillimage pkg/oci)   :2026-07, 2w
        Marketplace/repo management               :2026-07, 2w
    section Release
        Feature parity validation                  :2026-07, 1w
        GoReleaser setup and first release         :2026-07, 1w

>         Remove Python source                       :2026-08, 1w

Reviewer: We should update the dates here. Let's see if we can set a realistic start date.

Author: Adjusted all

```

## Coexistence Layout

During the transition, both Python and Go source live in the same repository:

```
lola/
├── src/lola/              # FROZEN — Python source, tagged v1.x-python-final
├── cmd/lola/main.go       # NEW — Go entry point
├── internal/              # NEW — Go private packages
├── pkg/                   # NEW — Go public packages
├── go.mod                 # NEW
├── go.sum                 # NEW
├── .goreleaser.yaml       # NEW
├── pyproject.toml         # FROZEN — Python build config
├── uv.lock                # FROZEN
└── tests/                 # Existing Python tests (frozen)
```

CI runs both test suites:
- Python: `pytest` (existing, frozen)
- Go: `go test ./...` (new, growing)

## Tech Stack Details

### Cobra Command Registration

Each command is a separate file in `internal/cli/` following the one-file-per-command pattern. The root command registers all subcommands explicitly:

```go
func NewRootCmd() *cobra.Command {
    root := &cobra.Command{Use: "lola"}
    root.AddCommand(
        NewModCmd(),
        NewSkillCmd(),
        NewPluginCmd(),

>         NewGroupCmd(),

Reviewer: If "Group" is just a special type of empty module should it have it's own command tree?

Author: Yes it does, the idea is similar to `dnf group` where you can install a named bundle of related skills together. Updated the example to annotate it as a new Q3 command, check it out.

        NewRepoCmd(),
        NewExtCmd(),
        NewInstallCmd(),
        NewUpdateCmd(),
        NewSearchCmd(),

>         NewServeCmd(),

Reviewer: What does serve do?

Author: Serve is a placeholder to extend to an HTTP router. The idea is to have under the serve, either API support and repository serving. However, I haven't clearly defined this in the architecture; I left it as a placeholder for triggering this discussion. I also updated, adding gin-gonic in the package list. But again, not a factual design for it, just a gut feeling that we're gonna need this at some point.

>         NewServeCmd(),

Reviewer 2: The python CLI has: `mod`, `market`, `install`, `uninstall`, `update`, `list`, `sync`, `completions`. This example registers `SkillCmd`, `PluginCmd`, `GroupCmd`, `ExtCmd`, and `ServeCmd` which don't exist yet.

The ADR says "all existing commands retain identical behavior" but this mixes migration scope with new feature scope without calling it out. Worth annotating which ones are new vs migrated, or splitting the example to show migrated commands first and new ones separately.

Author: Fixed, separated migrated commands from new ones with clear annotations, added the missing `uninstall`, `list`, `sync` and `completions`, and removed the undocumented ones. Check the latest and feel free to close if you're good.

Viper handles all configuration sources in priority order:
1. CLI flags (highest)
2. Environment variables (`LOLA_*` prefix)
3. Project config (`lola.yml` or `lola.toml`)
4. Global config (`~/.lola/config.yml`)
5. Defaults (lowest)

Viper brings YAML (`go.yaml.in/yaml/v3`) and TOML (`pelletier/go-toml/v2`) as its own transitive dependencies — no separate imports needed.

### skillimage Package Usage

```go
import (
    "github.com/redhat-et/skillimage/pkg/oci"
    "github.com/redhat-et/skillimage/pkg/skillcard"
    "github.com/redhat-et/skillimage/pkg/lifecycle"
)
```

The OCI source handler uses `pkg/oci` to pull and unpack skill images, `pkg/skillcard` to parse `skill.yaml` metadata alongside `SKILL.md` frontmatter, and `pkg/lifecycle` to check skill lifecycle state (warn before installing deprecated skills).

### Frontmatter Parser

Hand-rolled, no external dependency:

```go
func ParseFrontmatter(content []byte, v any) (body []byte, err error) {
    parts := bytes.SplitN(content, []byte("---"), 3)
    if len(parts) < 3 {
        return content, nil
    }
    if err := yaml.Unmarshal(parts[1], v); err != nil {
        return nil, fmt.Errorf("parsing frontmatter: %w", err)

>     }

Reviewer 2: This will silently misparse files that don't start with `---`. If a markdown file has a horizontal rule mid-document:
```
Some introduction text.

---
description: not actually frontmatter
---

Body here.
```
`SplitN` produces 3 parts, so the function parses the middle section as YAML and drops everything before the first `---`. The Python `python-frontmatter` library requires the opening `---` at byte 0 and the existing test `test_parse_without_frontmatter` validates this.

We need a guard that `parts[0]` is empty:
```
if len(parts) < 3 || len(bytes.TrimSpace(parts[0])) > 0 {
    return content, nil
}
```
Also, `bytes.TrimLeft(parts[2], "\n")` strips all leading newlines, not just the one after the closing `---`. `bytes.TrimPrefix` with a single `"\n"` would preserve intentional blank lines at the start of the body.

Author: Nice catch, fixed both, added the guard for non-empty `parts[0]` and switched to `bytes.TrimPrefix` so we only strip the single trailing newline. Feel free to close if you're good.

    return bytes.TrimLeft(parts[2], "\n"), nil
}
```

## Python Feature Parity Checklist

> Before removing Python source, the Go binary must pass:

Reviewer 2: This checklist is missing a few things that exist in the current python CLI:

- lola uninstall and lola list commands
- user-scope vs project-scope installation (the `--scope` user flag)
- shell completions generation (lola completions bash/zsh/fish)
- managed-section target pattern used by Gemini and OpenCode
- `lola-module/` content directory convention
- version specifier parsing in sync (tilde, caret, semver ranges)
- `content_dirname` persistence in `source.yml`
- `append_context` on Installation records
- interactive prompts for conflict resolution

If this checklist is the gate for removing Python source, anything not on it regresses silently. Might be worth generating it from lola --help output and the test file names to make sure nothing slips through.

Author: Fixed, expanded the checklist with all the missing items you pointed out — `lola uninstall`, `lola list`, `--scope user`, completions, managed-section target, `lola-module/` convention, `content_dirname`, `append_context`, interactive prompts and version specifiers in sync. If we are missing anything else feel free to add.

└── tests/                 # Existing Python tests (frozen)
```

CI runs both test suites:
- Python: `pytest` (existing, frozen)
- Go: `go test ./...` (new, growing)

## Tech Stack Details

### Cobra Command Registration

Each command is a separate file in `internal/cli/` following the one-file-per-command pattern. The root command registers all subcommands explicitly:

```go
func NewRootCmd() *cobra.Command {
    root := &cobra.Command{Use: "lola"}
    root.AddCommand(
        // Migrated from Python CLI — target: Q2 end (2026-07)
        NewModCmd(),         // lola mod (add, rm, ls, info, update, search, init)
        NewInstallCmd(),     // lola install
        NewUninstallCmd(),   // lola uninstall
        NewUpdateCmd(),      // lola update
        NewListCmd(),        // lola list
        NewSyncCmd(),        // lola sync
        NewRepoCmd(),        // lola repo (renamed from market)
        NewCompletionsCmd(), // lola completions bash/zsh/fish

        // New commands — target: Q3 start (2026-08)
        NewSkillCmd(),  // lola skill — standalone skill management
        NewGroupCmd(),  // lola group — group install (dnf group-style bundles)

>         NewServeCmd(),  // lola serve — REST API and local repo server (gin-gonic)

Reviewer: Let's delay this feature till after we go to golang. I want to think on this more.

Author: Agreed, removed ServeCmd from the parity scope, it has no equivalent in the current Python CLI and the HTTP router design needs more thought. Keeping the checklist focused on 1:1 parity with Python. Serve can get its own design doc when we are ready.

- [ ] `lola mod search <query>` — across all enabled repos
- [ ] `lola install [MODULE] [-a ASSISTANT] [-f] [-v] [--scope project|user] [--append-context PATH] [--pre-install SCRIPT] [--post-install SCRIPT] [--workspace NAME] [PROJECT_PATH]`
- [ ] `lola uninstall <module> [-a ASSISTANT] [--scope project|user] [-v] [PROJECT_PATH]`
- [ ] `lola update [MODULE] [-a ASSISTANT] [-v]` — with orphan detection and removal for skills, commands, agents, and MCPs
- [ ] `lola list [--assistant ASSISTANT]`
- [ ] `lola sync [PROJECT_PATH] [-a ASSISTANT] [--dry-run] [-v]`
- [ ] `lola repo add/ls/set/rm/update` (renamed from `market`)
- [ ] `lola completions <bash|zsh|fish>` — shell completion scripts
- [ ] Shell completion for module names on `lola install` / `lola uninstall`

**Targets**

- [ ] All 5 targets: `claude-code`, `cursor`, `gemini-cli`, `openclaw`, `opencode`
- [ ] File-based target pattern (Claude Code, Cursor, OpenClaw) — skills, commands, agents written as individual files
- [ ] Managed-section target pattern (Gemini CLI, OpenCode) — skills appended to / removed from named sections inside `GEMINI.md` / `AGENTS.md`

> - [ ] OpenClaw `--workspace` flag (named workspace directories)

Reviewer: Can we rename `--workspace` to something more assistant-agnostic? I can imagine this being useful for other assistants too, not just OpenClaw.

What if it were `--install-root`?

For example:

`lola install <name> -a <assistant> --scope user --install-root=/path/to/custom/root`

That would let each assistant resolve its normal files relative to a custom install root.

Author: We can have this as a global flag for sure. But I still believe that when we support targets as "extensions," the flags to that command will not be global. 

The install root is actually already there, you can do `lola install my-mod ./some/path` and that positional path already works as the root, no flag needed. For `--workspace` that one stays tied to openclaw only, openclaw workspaces are not just a path, there is more to it and keeping it as a target specific flag is the right separation of concerns. We can expand on it post parity within the openclaw target.

**Module handling**

- [ ] Module content auto-discovery: `module/` → `lola-module/` → repo root (checked in order)

> - [ ] `--module-content` override and `content_dirname` persistence in `source.yml`

Reviewer: I also wonder if `--module-content` should be renamed to describe that it is a path within the source.

Maybe `--content-path`?

For example:

`lola mod add https://github.com/company/monorepo.git --content-path=packages/lola-skills`

That reads to me as “fetch this source, then load the Lola module content from this subpath.” It also keeps `/` intuitive for root content:

`lola mod add https://github.com/user/flat-repo.git --content-path=/`

Author: Agree, the name is better. But this is not set in stone, the idea of this document is just to track the Go migration for parity, the goal is 1:1 with current Python behavior, so we're keeping `--module-content` and also documenting the `--content-path` as an alias, so we do not break current behavior. Adjusted in the docs

- [ ] Single-skill layout (`SKILL.md` at root) and skill-pack layout (`skills/<name>/SKILL.md`)
- [ ] `AGENTS.md`, `commands/`, `agents/`, `mcps.json` auto-discovery within content dir
- [ ] `append_context` flag on install; persisted on `Installation` record; honoured on `lola update`
- [ ] Installation scope (`project` / `user`) on install, uninstall, and list

**Install behaviour**

> - [ ] Pre/post install hook execution (`lola.yaml` hooks and `--pre-install` / `--post-install` overrides)

Reviewer: We should detect both `lola.yml` and `lola.yaml`.

Author: Good catch, added to the parity checklist — Go will detect both `lola.yml` and `lola.yaml`. Check the latest.

- [ ] Interactive prompts: overwrite confirmation, multi-repo conflict picker (when module exists in more than one enabled repo)

> - [ ] Skill name conflict resolution: when two installed modules share a skill name, the second is installed with a `<module>.<skill>` prefix to avoid overwriting the first

Reviewer: Well... If there is a conflict we should probably prompt the user (asking them what prefix to use) or require a `--prefix` flag. We could also add `prefix: "myprefix_"` to `lola.yaml`?

Author: For the parity goal we are replicating the current Python behavior, silent prefixing when there is a conflict. Your suggestions around prompting the user, a `--prefix` flag or `prefix:` in `lola.yaml` are really good ideas and we do not want to lose them, so I added them to a Post-Parity Improvements section in the design doc so we can track those for the first improvement cycle after Go reaches parity. Check the latest.

> - [ ] Backwards-compatible uninstall: also removes old prefixed filenames (`<module>.<cmd>.md`, `<module>.<agent>.md`) left by installs made before the prefix was removed

Reviewer: Hmmm. This seems like it would be bug prone. The contract should probably be:

> Backwards compatibility is broken with the golang port. It will be a different major version. Uninstall should ONLY remove the files that were installed by `lola install`. This requires explicitly tracking the name by which we installed the skills, commands, etc.

Author: Indeed, it can be, but it's a current behavior that we are looking for parity. My recommendation here is to keep til we have parity. Once we achieve and tag, we can change.

> - [ ] Auto-search enabled repos when a module is not found in the local registry — equivalent to `_fetch_from_marketplace()` in the Python CLI

Reviewer: Shouldn't search be universal? So:

```suggestion
- [ ] Search works across the local module registry and any enabled repositories.
```

**Sync**

> - [ ] `.lola-req` parsing with PEP 440 version specifiers (`>=`, `~=` tilde, `^` caret, semver ranges)

Reviewer: Yeah this is a good callout. We should be referring to [Version Specifiers](https://packaging.python.org/en/latest/specifications/version-specifiers/#version-specifiers) instead.

> - [ ] `.lola-req` parsing with PEP 440 version specifiers (`>=`, `~=` tilde, `^` caret, semver ranges)

Author: Fixed, updated the checklist to reference the Python Version Specifiers spec and separated the Lola-specific `~` tilde and `^` caret extensions into their own item. Check the latest.
